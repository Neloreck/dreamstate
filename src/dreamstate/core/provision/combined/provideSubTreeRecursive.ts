import { createElement, ReactNode } from "react";

import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Recursively renders a tree of context providers and wraps the provided component with them.
 *
 * This function generates a React subtree by recursively creating provider components for each context manager
 * in the `sources` array. Each provider in the tree passes down its data to the children components, ensuring that
 * the component tree is properly provided with the necessary context values.
 *
 * @param {ReactNode} bottom - The component (children) that will be wrapped by the context providers.
 * @param {Array<TAnyContextManagerConstructor>} sources - An array of context manager class references that will
 *   provide data via context providers in the React tree.
 * @param {Map<TAnyContextManagerConstructor, TAnyObject>} contextStateRegistry - A map that holds the current
 *   state of each context manager, which will be provided to the corresponding context provider.
 * @param {number} current - The current iteration index for the recursive provider creation process.
 * @returns {ReactNode} A React node representing the tree of context providers, wrapped around the `bottom` component.
 */
export function provideSubTreeRecursive(
  bottom: ReactNode = null,
  sources: Array<TAnyContextManagerConstructor>,
  contextStateRegistry: Map<TAnyContextManagerConstructor, TAnyObject>,
  current: number = 0
): ReactNode {
  /*
   * Wrap children node with providers until all source provider elements are created.
   * Get latest state for provision from current scope registry.
   */
  return current >= sources.length
    ? bottom
    : createElement(
      sources[current].REACT_CONTEXT.Provider,
      { value: contextStateRegistry.get(sources[current]) },
      provideSubTreeRecursive(bottom, sources, contextStateRegistry, current + 1)
    );
}
