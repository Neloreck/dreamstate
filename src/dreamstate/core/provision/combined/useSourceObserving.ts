import { useContext, useEffect, useMemo } from "react";

import { log } from "@/macroses/log.macro";

import { IScopeContext, ScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { useForceUpdate } from "@/dreamstate/core/utils/useForceUpdate";
import { TAnyContextManagerConstructor, TAnyObject, TCallable } from "@/dreamstate/types";

/**
 * A hook used internally for observing and injecting data based on the current scope.
 *
 * This hook consumes the provided context manager classes (`sources`) and the initial state,
 * performing data injection and provisioning within the context of the current scope. It returns
 * the current state registry, which is used for data provisioning throughout the component tree.
 *
 * @param {Array<TAnyContextManagerConstructor>} sources - An array of context manager class references
 *   that will be observed for state changes.
 * @param {TAnyObject | undefined} initialState - The initial state to be used for data provisioning
 *   when the hook is first invoked.
 * @returns {Map<TAnyContextManagerConstructor, TAnyObject>} A map of context manager classes and their
 *   corresponding current states, used for providing data to components.
 */
export function useSourceObserving(
  sources: Array<TAnyContextManagerConstructor>,
  initialState: TAnyObject | undefined
): Map<TAnyContextManagerConstructor, TAnyObject> {
  const scope: IScopeContext = useContext(ScopeContext);
  const forceUpdate: TCallable = useForceUpdate();

  /*
   * Warn if current observer is mounted out of scope in dev mode.
   */
  if (IS_DEV) {
    if (!scope) {
      log.error("Dreamstate providers should be used in a scope. Wrap your component tree with ScopeProvider");
    }
  }

  /*
   * Use memo for first and single init of required components.
   * The point is to force registering before provider rendering for first init.
   *
   * Registering will work in the same way even if it is called multiple times.
   * Dependencies array is mostly used for HMR updates to force reloading on class reference changes.
   */
  useMemo(function(): void {
    for (let it = 0; it < sources.length; it ++) {
      scope.INTERNAL.registerManager(sources[it], initialState);
    }
  }, sources);

  /*
   * Mount current observers and trigger related lifecycle methods when needed.
   * Count references of providers to detect whether we start provisioning or ending it.
   *
   * ! Iteration order is related to actual tree mount-unmount order imitating.
   * ! Dependencies array is mostly used for HMR updates to force reloading on class reference changes.
   */
  useEffect(function(): TCallable {
    let isSomethingRegistered: boolean = false;

    for (let it = sources.length - 1; it >= 0; it --) {
      /*
       * If something is registered, memoize it and force update to sync useEffect and actual provision tree.
       */
      const registered: boolean = scope.INTERNAL.registerManager(sources[it], initialState);

      scope.INTERNAL.addServiceObserver(sources[it], forceUpdate);

      /*
       * Should stick to 'false' until at least one result is 'true'.
       */
      isSomethingRegistered = registered || isSomethingRegistered;
    }

    /*
     * Re-sync scope providers if something was registered.
     * Normally happens with HMR chunks exchange and caused problems that use react context without subscription.
     * Required to force render of subscribed components after HMR with newest fast-refresh plugins.
     */
    if (isSomethingRegistered) {
      forceUpdate();
    }

    return function(): void {
      for (let it = 0; it < sources.length; it ++) {
        scope.INTERNAL.removeServiceObserver(sources[it], forceUpdate);
      }
    };
  }, sources);

  return scope.INTERNAL.REGISTRY.CONTEXT_STATES_REGISTRY;
}
