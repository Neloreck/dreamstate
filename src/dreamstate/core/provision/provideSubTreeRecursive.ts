import { createElement, ReactElement } from "react";

import { CONTEXT_STATES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function provideSubTreeRecursive(
  bottom: ReactElement,
  sources: Array<TAnyContextManagerConstructor>,
  current: number
): ReactElement {
  return current >= sources.length
    ? bottom
    : createElement(
      sources[current].REACT_CONTEXT.Provider,
      { value: CONTEXT_STATES_REGISTRY.get(sources[current]) },
      provideSubTreeRecursive(bottom, sources, current + 1)
    );
}
