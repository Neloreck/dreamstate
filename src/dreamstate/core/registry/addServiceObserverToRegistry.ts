import { CONTEXT_OBSERVERS_REGISTRY } from "@/dreamstate/core/internals";
import { TDreamstateService, TUpdateObserver } from "@/dreamstate/types";

/**
 * Add state changes observer.
 */
export function addServiceObserverToRegistry(
  Service: TDreamstateService,
  observer: TUpdateObserver
): void {
  CONTEXT_OBSERVERS_REGISTRY.get(Service)!.add(observer);
}
