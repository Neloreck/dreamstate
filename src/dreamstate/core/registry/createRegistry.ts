import {
  TAnyContextManagerConstructor, TAnyObject,
  TQueryListener,
  TQueryType,
  TServiceMap,
  TSignalListener, TUpdateObserver, TUpdateSubscriber
} from "@/dreamstate/types";

export interface IRegistry {
  SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>>;
  QUERY_PROVIDERS_REGISTRY: Map<TQueryType, Array<TQueryListener<any, any>>>;
  CONTEXT_SERVICES_REGISTRY: TServiceMap<InstanceType<TAnyContextManagerConstructor>>;
  CONTEXT_STATES_REGISTRY: TServiceMap<TAnyObject>;
  CONTEXT_SERVICES_REFERENCES: TServiceMap<number>;
  CONTEXT_SERVICES_ACTIVATED: Set<TAnyContextManagerConstructor>;
  // ContextManagers observers registry - Providers of related context.
  CONTEXT_OBSERVERS_REGISTRY: TServiceMap<Set<TUpdateObserver>>;
  // ContextManagers subscribers registry for data updates.
  CONTEXT_SUBSCRIBERS_REGISTRY: TServiceMap<Set<TUpdateSubscriber<any>>>;
}

/**
 * Create registry with internal stores.
 */
export function createRegistry(): IRegistry {
  return ({
    CONTEXT_SERVICES_ACTIVATED: new Set(),
    QUERY_PROVIDERS_REGISTRY: new Map(),
    SIGNAL_LISTENERS_REGISTRY: new Set(),
    CONTEXT_SERVICES_REGISTRY: new Map(),
    CONTEXT_STATES_REGISTRY: new Map(),
    CONTEXT_SERVICES_REFERENCES: new Map(),
    CONTEXT_OBSERVERS_REGISTRY: new Map(),
    CONTEXT_SUBSCRIBERS_REGISTRY: new Map()
  });
}
