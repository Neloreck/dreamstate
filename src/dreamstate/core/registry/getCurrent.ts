import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyObject, TDreamstateService } from "@/dreamstate/types";

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TDreamstateService<any>>(
  Service: T
): InstanceType<T> | null {
  return CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T> || null;
}
