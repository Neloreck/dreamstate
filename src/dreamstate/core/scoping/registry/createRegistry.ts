import {
  TAnyObject,
  TQueryListener,
  TQueryType,
  TManagerInstanceMap,
  TManagerMap,
  TSignalListener,
  TUpdateObserver,
  TUpdateSubscriber
} from "@/dreamstate/types";

/**
 * Internal registry object that stores mutable references accounting all dreamstate scope data.
 */
export interface IRegistry {
  /**
   * Registry of currently provided context managers instances.
   */
  CONTEXT_INSTANCES_REGISTRY: TManagerInstanceMap;
  /**
   * Registry of current scope manager states.
   * Used for data accessing and sync across providers.
   */
  CONTEXT_STATES_REGISTRY: TManagerMap<TAnyObject>;
  /**
   * Provider references counting.
   * Used to determine whether specific manager should or should not exist.
   * If at least one provider exists, context manager should exist in current scope.
   * If reference count is 0, context manager instance should be disposed.
   */
  CONTEXT_SERVICES_REFERENCES: TManagerMap<number>;
  /**
   * ContextManagers observers registry for Providers of related context.
   * Used by dreamstate observer components.
   */
  CONTEXT_OBSERVERS_REGISTRY: TManagerMap<Set<TUpdateObserver>>;
  /**
   * ContextManagers subscribers registry for data updates.
   * Used by pub-sub subscribers that consume data directly.
   */
  CONTEXT_SUBSCRIBERS_REGISTRY: TManagerMap<Set<TUpdateSubscriber<any>>>;
  /**
   * Registry of signal handlers.
   */
  SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any>>;
  /**
   * Registry of custom query requests handlers.
   */
  QUERY_PROVIDERS_REGISTRY: Map<TQueryType, Array<TQueryListener<any, any>>>;
}

/**
 * Create registry with internal stores.
 * Used in current scope to store mutable data.
 */
export function createRegistry(): IRegistry {
  return {
    QUERY_PROVIDERS_REGISTRY: new Map(),
    SIGNAL_LISTENERS_REGISTRY: new Set(),
    CONTEXT_INSTANCES_REGISTRY: new Map(),
    CONTEXT_STATES_REGISTRY: new Map(),
    CONTEXT_SERVICES_REFERENCES: new Map(),
    CONTEXT_OBSERVERS_REGISTRY: new Map(),
    CONTEXT_SUBSCRIBERS_REGISTRY: new Map()
  };
}
