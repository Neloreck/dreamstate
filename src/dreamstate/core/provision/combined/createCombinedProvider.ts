import { FunctionComponent, ReactNode } from "react";

import { provideSubTreeRecursive } from "@/dreamstate/core/provision/combined/provideSubTreeRecursive";
import { useSourceObserving } from "@/dreamstate/core/provision/combined/useSourceObserving";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";
import { IProviderProps } from "@/dreamstate/types/provision";

/**
 * Creates a provider that unifies multiple data sources and attempts to re-render all providers on every update.
 *
 * This function generates a React provider component that observes changes across all specified context
 * managers. This combined provider triggers a re-render for all related provider components whenever any single
 * manager's state is updated.
 *
 * For example, if you have 10 sources, all 10 context provider components will re-render when any one of them
 * updates, and the difference between the current and previous state is matched using React's reference checks.
 *
 * @param {Array<TAnyContextManagerConstructor>} sources - An array of context manager class references
 *   that should be combined into a single provider.
 * @returns {FunctionComponent} A React function component that acts as a combined provider for the specified
 *   context manager classes, ensuring synchronized re-renders across all providers.
 */
export function createCombinedProvider<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: IProviderProps<T>): ReactNode {
    const registry: Map<TAnyContextManagerConstructor, TAnyObject> = useSourceObserving(sources, props.initialState);

    return provideSubTreeRecursive(props.children, sources, registry);
  }

  /*
   * One observer for all provider sources and many context providers.
   */
  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextManagerConstructor): string {
      return it.name;
    })}]`;
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
