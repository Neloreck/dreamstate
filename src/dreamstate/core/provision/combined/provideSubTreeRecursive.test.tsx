import { shallow } from "enzyme";
import { default as React } from "react";

import { provideSubTreeRecursive } from "@/dreamstate/core/provision/combined/provideSubTreeRecursive";
import { createRegistry, IRegistry } from "@/dreamstate/core/scoping/registry/createRegistry";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";
import { NestedManager, TestManager } from "@/fixtures";

describe("provideSubTreeRecursive rendering", () => {
  it("Should match snapshot", () => {
    const registry: IRegistry = createRegistry();
    const providers: Array<TAnyContextManagerConstructor> = [
      TestManager,
      NestedManager,
      TestManager
    ];

    const emptyRecursiveRender = shallow(
      <div>
        {
          provideSubTreeRecursive(
            <div> bottom </div>,
            providers,
            registry.CONTEXT_STATES_REGISTRY
          )
        }
      </div>
    );

    expect(emptyRecursiveRender).toMatchSnapshot();

    registry.CONTEXT_STATES_REGISTRY.set(TestManager, { first: "is" });
    registry.CONTEXT_STATES_REGISTRY.set(NestedManager, { second: "is" });

    const providedRecursiveRender = shallow(
      <div>
        {
          provideSubTreeRecursive(
            <div> bottom </div>,
            providers,
            registry.CONTEXT_STATES_REGISTRY
          )
        }
      </div>
    );

    expect(providedRecursiveRender).toMatchSnapshot();
  });
});
