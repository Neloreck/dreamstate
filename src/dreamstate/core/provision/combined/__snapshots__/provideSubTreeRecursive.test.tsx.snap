// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provideSubTreeRecursive rendering should match snapshot 1`] = `
<div>
  <ContextProvider>
    <ContextProvider>
      <ContextProvider>
        <div>
           bottom 
        </div>
      </ContextProvider>
    </ContextProvider>
  </ContextProvider>
</div>
`;

exports[`provideSubTreeRecursive rendering should match snapshot 2`] = `
<div>
  <ContextProvider
    value={
      Object {
        "first": "is",
      }
    }
  >
    <ContextProvider
      value={
        Object {
          "second": "is",
        }
      }
    >
      <ContextProvider
        value={
          Object {
            "first": "is",
          }
        }
      >
        <div>
           bottom 
        </div>
      </ContextProvider>
    </ContextProvider>
  </ContextProvider>
</div>
`;
